Source/sdc/sdc_util.cpp

```diff
@@ -0,0 +1,429 @@
+        {
+            Real integral = (dt/dt_m) * (1.0_rt/600.0_rt) *
+                ((168.0_rt - 73.0_rt*sqrt(6.0_rt))*(A_1_old(i,j,k,n) + R_1_old(i,j,k,n)) +
+                 (120.0_rt + 5.0_rt*sqrt(6.0_rt))*(A_2_old(i,j,k,n) + R_2_old(i,j,k,n)) +
+                 (72.0_rt + 8.0_rt*sqrt(6.0_rt))*(A_3_old(i,j,k,n) + R_3_old(i,j,k,n)));
+
+            C(i,j,k,n) = (A_m(i,j,k,n) - A_2_old(i,j,k,n)) - R_3_old(i,j,k,n) + integral;
+        });
+    }
+    else
+    {
+        amrex::Abort("error in ca_sdc_compute_C4 -- should not be here");
+    }
+} //  end ca_sdc_compute_C4_radau
+
+void
+Castro::ca_sdc_conservative_update(const amrex::Box& bx, amrex::Real const dt_m,
+                                   amrex::Array4<const amrex::Real> const& U_old,
+                                   amrex::Array4<amrex::Real> const& U_new,
+                                   amrex::Array4<const amrex::Real> const& C,
+                                   amrex::Array4<const amrex::Real> const& R_new)
+{
+    // given <U>_old, <R>_new, and <C>, compute <U>_new
+
+    // now consider the reacting system
+    AMREX_PARALLEL_FOR_4D(bx, U_new.nComp(), i, j, k, n,
+    {
+        U_new(i,j,k,n) = U_old(i,j,k,n) + dt_m * R_new(i,j,k,n) + dt_m * C(i,j,k,n);
+    });
+} // end subroutine ca_sdc_conservative_update
+#endif
+
+
+void Castro::ca_sdc_compute_initial_guess(const amrex::Box& bx,
+                                          amrex::Array4<const amrex::Real> const& U_old,
+                                          amrex::Array4<const amrex::Real> const& U_new,
+                                          amrex::Array4<const amrex::Real> const& A_old,
+                                          amrex::Array4<const amrex::Real> const& R_old,
+                                          amrex::Array4<amrex::Real> const& U_guess,
+                                          amrex::Real const dt_m, int const sdc_iteration)
+{
+    // compute the initial guess for the Newton solve
+    // Here dt_m is the timestep to update from time node m to m+1
+
+    if (sdc_iteration == 0)
+    {
+        AMREX_PARALLEL_FOR_4D(bx, U_guess.nComp(), i, j, k, n,
+        {
+            U_guess(i,j,k,n) = U_old(i,j,k,n) + dt_m * A_old(i,j,k,n) + dt_m * R_old(i,j,k,n);
+        });
+    }
+    else
+    {
+        AMREX_PARALLEL_FOR_4D(bx, U_guess.nComp(), i, j, k, n,
+        {
+            U_guess(i,j,k,n) = U_new(i,j,k,n);
+        });
+    }
+
+}
+
+
+#ifdef REACTIONS
+void Castro::ca_store_reaction_state(const amrex::Box& bx,
+                                     amrex::Array4<const amrex::Real> const& R_old,
+                                     amrex::Array4<const amrex::Real> const& state,
+                                     amrex::Array4<amrex::Real> const& R_store)
+{
+    // copy the data from the last node's reactive source to the state data
+
+    // for R_store we use the indices defined in Castro_setup.cpp for
+    // Reactions_Type
+    int nspec = R_store.nComp()-2;
```

Comment:
it is safer to use `NumSpec` from the `network.H`

Reply:
It looks like that's a `constexpr int`, which should be usable inside the lambda. Is it cleaner to just use `NumSpec` directly, or should I keep using the locally scoped copy?

Reply:
I think you should just use NumSpec directly
that's what we do elsewhere in the code.